From de90c3d852320e4303b28593e068fe79008ba383 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ball=C3=B3=20Gy=C3=B6rgy?= <ballogyor@gmail.com>
Date: Sat, 5 Apr 2025 15:21:12 +0200
Subject: [PATCH] Add "new-window" command-line option

And add back the action to the desktop file, so it will work not only in
GNOME Shell, but also in KDE and other environments that support actions.
---
 data/org.gnome.Gtranslator.desktop.in.in |  7 +++++++
 src/gtr-application.c                    | 10 ++++++++++
 2 files changed, 17 insertions(+)

diff --git a/data/org.gnome.Gtranslator.desktop.in.in b/data/org.gnome.Gtranslator.desktop.in.in
index 44213468b..f7eaaf735 100644
--- a/data/org.gnome.Gtranslator.desktop.in.in
+++ b/data/org.gnome.Gtranslator.desktop.in.in
@@ -11,3 +11,10 @@ StartupNotify=true
 MimeType=text/x-gettext-translation;text/x-gettext-translation-template;
 Categories=GNOME;GTK;Development;Utility;Translation;
 Keywords=Gtranslator;Translation;Editor;Gettext;Write;
+Actions=new-window;
+
+[Desktop Action new-window]
+Exec=gtranslator --new-window
+Name=New Window
+# Translators: Do NOT translate or transliterate this text (this is an icon file name)!
+Icon=window-new
diff --git a/src/gtr-application.c b/src/gtr-application.c
index 3b87f882a..f12b495aa 100644
--- a/src/gtr-application.c
+++ b/src/gtr-application.c
@@ -118,6 +118,15 @@ on_window_delete_event_cb (GtrWindow * window,
 
 static int
 handle_local_options_cb (GApplication *application, GVariantDict *options, gpointer user_data) {
+  if (g_variant_dict_contains (options, "new-window")) {
+    g_application_register (application, NULL, NULL);
+
+    if (g_application_get_is_remote (application)) {
+      g_action_group_activate_action (G_ACTION_GROUP (application), "new-window", NULL);
+      return 0;
+    }
+  }
+
   if (g_variant_dict_contains (options, "version")) {
     g_print ("%s - %s\n", PACKAGE, PACKAGE_VERSION);
     return 0;
@@ -136,6 +145,7 @@ gtr_application_init (GtrApplication *application)
   priv->first_run = FALSE;
 
   GOptionEntry options[] = {
+      {"new-window", 'w', 0, G_OPTION_ARG_NONE, NULL, "Open a new window", NULL},
       {"version", 'v', 0, G_OPTION_ARG_NONE, NULL, "Print version information and exit", NULL},
       {NULL}
   };
-- 
GitLab

